Removes GUI dependencies from the core class ConfigManager

michaelgregorius opened the pull request:
Adds a struct `ConfigManager::ConfigLoadResult` which holds the result of
the method `ConfigManager::loadConfigFile`. Doing it this way the caller
(`main.cpp`) can act accordingly in case there is an error. The current
implementation shows the warning message box that was shown before when
we are on the GUI code path.

`ConfigManager::saveConfigFile` now returns a boolean to indicate whether
the saving worked or not. Also in this case it's up to the caller to act
accordingly. In this case the sole caller `SetupDialog` shows the previous
critical error message. To be able to include the config file's name a new
method `ConfigManager::lmmsRcFile` was added.

Wallacoloo commented on include/ConfigManager.h:
```diff
@@ -230,8 +235,27 @@ class EXPORT ConfigManager
+		bool failuresOccurred;
+		int errorLine;
+		int errorColumn;
+		QString errorString;
```
Is everyone OK with this naming convention? At the moment, we seem to have two conflicting conventions for struct member variables:

We have the classic `m_varName` (e.g. https://github.com/LMMS/lmms/blob/master/include/MainWindow.h#L187) and also just `varName` (e.g. https://github.com/LMMS/lmms/blob/master/include/Mixer.h#L86).

Arguments can be made for either one. Personally, I've always felt that the `m_` denotes some amount of privacy, and that variables meant to be accessed externally shouldn't have the `m_`. The `m_` convention is really only useful when accessing a variable using an implicit `[this->]m_var` to emphasis where the variable lives.

So I prefer the naming @michaelgregorius has chosen here for Plain-Old-Data types, but we should establish a convention.

michaelgregorius replied:
@Wallacoloo That was exactly my reasoning. I consider member variables prefixed with `m_` private to the class and only accessible by getters and setters (if at all). I did not want to bloat the rather simple class `ConfigLoadResult` with setters and getters and calling something like `configLoadResult.m_failuresOccured` just seemed wrong. :)

Wallacoloo commented:
Looks pretty good. I left just one critique.

But can somebody help explain what the status of this is? I'm very confused looking at the recent events:
* #2536 is reported to have been merged into master, yet its changes do not appear to be present in current master. This PR contains both the commit found in #2536 as well as a new one.
* This PR is marked "in progress", which means to me that it's intended for more commits to be pushed here before merging. But the description leaves no indication that this isn't ready to be merged, or of what new commits are planned.

So first, wtf happened to #2536? If the merge was forceably rolled back, you **need** to leave a comment on the PR to explain that it was, and why.

Secondly, @Lukas-W @michaelgregorius is the PR ready to be merged after a reviewer approves it? As a reviewer, when I see "in progress", it tells me "oh, the author isn't done with this changeset yet. There's no reason for me to review it now, because the code is likely to change *significantly* before being ready to be merged".

If I'm understanding the "in progress" tag, I think its existence is a bit dubious, but it does at least allow people to see who's working on what so that we can avoid duplicated work. I have a few partially-completed changesets that I'm planning to finish in ~10 days or so. Should I be opening "in progress" PRs for them now, rather than waiting until they're polished & ready to be merged?

tresf commented:
> #2536 is reported to have been merged into master, yet its changes do not appear to be present in current master. This PR contains both the commit found in #2536 as well as a new one.

That's one of many disadvantages in forcing to master.

I didn't think the PR was even ready to be merged.  From my understanding, it was in preparation of better GUI/core separation without any functional differences, which IMO means it should be held until after `stable-1.2`, since it affects 113+ files.

But then we get into the conversation about maintaining a new `stable-1.2` branch, which I'm too busy to coordinate at the moment, hence my silence (until now I guess).

Lukas-W commented:
![image](https://cloud.githubusercontent.com/assets/2879917/12878960/605db2c0-ce91-11e5-9631-a60070282c1a.png)

I didn't even do that. Wtf?

michaelgregorius commented:
> Can you add a comment annotating the meaning of the boolean return value? (e.g. // returns true on success). There are alternative conventions in which non-zero return values represent errors instead, so I believe it should be made explicit.

@Wallacoloo I have added some documentation for the changed methods as well as for the new struct.

michaelgregorius commented:
> But can somebody help explain what the status of this is? I'm very confused looking at the recent events [...]

#2536 was about the first round of removals for the static calls into the class `Engine`. This pull request is more concerned with getting rid of the calls to the static variable `gui` which is a typedef to `GuiApplication::instance()`.

I think that getting rid of the `gui` calls is more important anyway as `gui` refers to an instance in the GUI layer which is definitively problematic when called from Core code whereas `Engine` is defined in the business anyway and should not be too much of a problem with regards to the separation (GUI is allowed to call business code and business is allowed to call business code). Although a clear separation (i.e. no more static instance of `Engine`) would also help to keep the code clean.

I have added a comment in #2536 which explains the situation.

michaelgregorius commented:
> Secondly, @Lukas-W @michaelgregorius is the PR ready to be merged after a reviewer approves it?

Seems like this is either a hiccup of GitHub or we have added some settings somewhere that the tag is set automatically for each new pull request. The tag was set immediately after I have opened the pull requests and I have now removed it.

An alternative explanation is that @Lukas-W is a cleverly disguised bot. :wink:

michaelgregorius commented:
> I didn't think the PR was even ready to be merged. From my understanding, it was in preparation of better GUI/core separation without any functional differences, which IMO means it should be held until after stable-1.2, since it affects 113+ files.

My reasoning was that although the changes should be low risk they would also get tested in the upcoming beta phase. I can also hold them until we have a `stable-1.2` branch and push them to `master` then.

> But then we get into the conversation about maintaining a new stable-1.2 branch, which I'm too busy to coordinate at the moment, hence my silence (until now I guess).

@tresf Are there any big blockers left with regards to a `stable-1.2` branch/release?

tresf commented:
> @tresf Are there any big blockers left with regards to a stable-1.2 branch/release?

None.  We're pretty much ready to start testing.

> My reasoning was that although the changes should be low risk they would also get tested in the upcoming beta phase.

Yeah, that's fine too.  Progress is good, just hard to testing a moving target.  I do believe the changes to be mostly benign, so I'm not concerned as much as I am cautious. :+1:

lukas-w commented:
As far as I can see, this PR seems to be a continuation of the `EngineClient` refactorings from #2536 (which has been reverted) and #2550 (which has been closed). Should we close this one as well?

michaelgregorius commented:
@lukas-w From my point of view we can close it. I have no intentions to merge / rebase this to the current master after such a long time anyway.
